### PR TITLE
[iOS] automaticallyAdjustKeyboardInsets not shifting scrollview content

### LINKED ISSUE #46595: automaticallyAdjustKeyboardInsets not shifting scrollview content
### Description

Looking at the explanation and examples of `automaticallyAdjustKeyboardInsets` here: https://github.com/facebook/react-native/pull/31402, I'd expect scrollview content to shift upwards when focusing an input in `InputAccessoryView`, but it doesn't. Instead the keyboard covers the bottom portion of the content and you need to manually scroll back up.

### Steps to reproduce

1. Load the snack
2. Focus the input
3. Observe the bottom portion of the list is covered

### React Native Version

0.75.3

### Affected Platforms

Runtime - iOS

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.4.1
  CPU: (12) arm64 Apple M2 Max
  Memory: 3.77 GB / 96.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.15.0
    path: ~/.nvm/versions/node/v20.15.0/bin/node
  Yarn: Not Found
  npm:
    version: 10.7.0
    path: ~/.nvm/versions/node/v20.15.0/bin/npm
  Watchman:
    version: 2024.07.08.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /Users/anthony.difrancesco/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK: Not Found
IDEs:
  Android Studio: Not Found
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 11.0.6
    path: /usr/bin/javac
  Ruby:
    version: 3.3.3
    path: /Users/anthony.difrancesco/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.75.3
    wanted: ~0.75.3
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: false
  newArchEnabled: false
```

### Stacktrace or Logs

```text
N/A
```

### Reproducer

https://snack.expo.dev/@acdifran/interactive-keyboard

### Screenshots and Videos

Observed|Expected
--|--
<video src="https://github.com/user-attachments/assets/5f0443df-2f72-4a9c-b024-1c0faaa9eb6a" width="320" height="240" controls></video>|<video src="https://github.com/user-attachments/assets/ea7968c3-53bb-455d-8f95-75738e9dc894" width="320" height="240" controls></video>

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary:

Fixes #46595 . It seems #37766 broke the `automaticallyAdjustKeyboardInsets` when input accessory view become first responder.

## Changelog:

[IOS] [FIXED] - automaticallyAdjustKeyboardInsets not shifting scrollview content

## Test Plan:

Repro please see in ##46595 .

### CODE REVIEW

File: packages/react-native/React/Views/UIResponder+React.m [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+{
+  reactCurrentFirstResponder = nil;
+  [[UIApplication sharedApplication] sendAction:@selector(reactFindFirstResponder:) to:nil from:nil forEvent:nil];
+  return reactCurrentFirstResponder;
+}
+
+- (void)reactFindFirstResponder:(id)sender
+{
+  reactCurrentFirstResponder = self;
+}
```
Comment (cipolleschi) [inline packages/react-native/React/Views/UIResponder+React.m:23]: This is something we used to have and we removed it. It is seen as an anti-pattern, so we'd rather not have it. Probably @javache has more context on why it got removed.

**Review by cipolleschi — CHANGES_REQUESTED** [review]

File: packages/react-native/React/Views/UIResponder+React.m (hunk 1 above)
Comment (zhongwuzw) [reply packages/react-native/React/Views/UIResponder+React.m]: @cipolleschi I changed another way to find first responder :)

Comment (cipolleschi) [reply packages/react-native/React/Views/UIResponder+React.m]: I found why: the previous API is a private API https://github.com/facebook/react-native/pull/37766#discussion_r1222735097

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm [hunk 2]
```diff
@@ -34,6 +34,7 @@ @implementation RCTBaseTextInputView {
 
 - (void)reactUpdateResponderOffsetForScrollView:(RCTScrollView *)scrollView
 {
+  scrollView.firstResponderFocusView = self.backedTextInputView;
```
Comment (javache) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm:37]: This should respect the early return.

File: packages/react-native/React/Views/ScrollView/RCTScrollView.h [hunk 3]
```diff
@@ -50,6 +50,8 @@
 @property (nonatomic, assign) BOOL inverted;
 /** Focus area of newly-activated text input relative to the window to compare against UIKeyboardFrameBegin/End */
 @property (nonatomic, assign) CGRect firstResponderFocus;
+/** newly-activated text input relative to the window to compare against UIKeyboardFrameBegin/End */
+@property (nonatomic, weak) UIView *firstResponderFocusView;
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.h:54]: There's a lot of overlap between these two fields. If we're going to be storing the focusView anyway, can we get rid of the focus rect and instead add a `reactFocusFrame` to `UIView+React`?

File: packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 4]
```diff
@@ -102,6 +102,7 @@ - (void)didMoveToWindow
 
 - (void)reactUpdateResponderOffsetForScrollView:(RCTScrollViewComponentView *)scrollView
 {
+  scrollView.firstResponderFocusView = _backedTextInputView;
```
Comment (javache) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:105]: Respect early return.

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm (hunk 2 above)
Comment (zhongwuzw) [reply packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm]: @javache I optimized the code to only store it when the first responder is outside the scroll view.

**Comment by zhongwuzw** [comment]
Hi @javache, just a friendly reminder— the issue reporter is asking for any updates regarding this matter. Could you please assist in resuming the review? Thank you!

File: packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.h [hunk 5]
```diff
@@ -38,6 +38,9 @@ NS_ASSUME_NONNULL_BEGIN
 /** Focus area of newly-activated text input relative to the window to compare against UIKeyboardFrameBegin/End */
 @property (nonatomic, assign) CGRect firstResponderFocus;
 
+/** newly-activated text input outside of the scroll view */
+@property (nonatomic, weak) UIView *firstResponderViewOffScrollView;
```
Comment (javache) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.h:42]: nit: firstResponderViewOutsideScrollView

File: packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm [hunk 6]
```diff
@@ -203,21 +204,27 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
...
-      return;
-    }
-
-    CGRect viewIntersection = CGRectIntersection(self.firstResponderFocus, keyboardEndFrame);
+      UIView *inputAccessoryView = _firstResponderViewOffScrollView.inputAccessoryView;
+      if (inputAccessoryView) {
+        // Text input view is within the inputAccessoryView.
+        contentDiff = keyboardEndFrame.origin.y - keyboardBeginFrame.origin.y;
+      } else {
+        // Text input view is outside of the scroll view.
```
Comment (javache) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm:212]: This is the case where there's **no** text input view found, right? Update comment?

File: packages/react-native/React/Views/ScrollView/RCTScrollView.m [hunk 7]
```diff
@@ -338,6 +338,12 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
     if (!didFocusExternalTextField && focusEnd > endFrame.origin.y) {
       // Text field active region is below visible area with keyboard - update diff to bring into view
       contentDiff = endFrame.origin.y - focusEnd;
+    } else {
+      UIView *inputAccessoryView = _firstResponderViewOffScrollView.inputAccessoryView;
+      if (inputAccessoryView) {
+        // Text input view is within the inputAccessoryView.
+        contentDiff = endFrame.origin.y - beginFrame.origin.y;
+      }
```
Comment (javache) [inline packages/react-native/React/Views/ScrollView/RCTScrollView.m:346]: I'm surprised the logic for Fabric and Paper has diverged here quite a bit - should this be unified?

File: packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm [hunk 8]
```diff
@@ -203,21 +204,24 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
...
-      // Text input view is outside of the scroll view.
-      return;
-    }
-
-    CGRect viewIntersection = CGRectIntersection(self.firstResponderFocus, keyboardEndFrame);
+      UIView *inputAccessoryView = _firstResponderViewOutsideScrollView.inputAccessoryView;
+      if (inputAccessoryView) {
+        // Text input view is within the inputAccessoryView.
+        contentDiff = keyboardEndFrame.origin.y - keyboardBeginFrame.origin.y;
+      }
```
Comment (javache) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm:211]: Do we still want an early return here? Or should we have a general check for `contentDiff == 0`?

Comment (zhongwuzw) [reply packages/react-native/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm:211]: I prefer to maintain the same logic as the old architecture, which does not perform early return checks or check for `contentDiff == 0`.

**Pull request closed without merging**